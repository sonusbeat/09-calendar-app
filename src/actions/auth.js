import types from '../types/types';
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import Swal from 'sweetalert2';


export const startLogin = ( email, password ) => {
  return async ( dispatch ) => {
    const response = await fetchWithoutToken( "auth", { email, password }, "POST" );
    const body = await response.json();

    if ( body.ok ) {
      // Se guarda el token en el storage
      localStorage.setItem('token', body.token);

      // Para saber cuando va a expirar el token
      localStorage.setItem( 'token-init-date', new Date().getTime() );

      // Dispara la acción del login
      dispatch(
        login({
          uid: body.uid,
          name: body.name
        })
      );
    } else {
      // En caso de que body.ok sea null,
      // lanza un mensaje de error
      Swal.fire("Error", body.msg, "error");
    }

  };
};

export const startRegister = ( name, email, password ) => {
  return async ( dispatch ) => {
    const response = await fetchWithoutToken( "auth/new", { name, email, password }, "POST" );
    const body = await response.json();

    if( body.ok ) {
      // Se guarda el token en el storage
      localStorage.setItem('token', body.token);

      // Para saber cuando va a expirar el token
      localStorage.setItem( 'token-init-date', new Date().getTime() );

      // Dispara la acción del login
      dispatch(
        login({
          uid: body.uid,
          name: body.name
        })
      );
    } else {
      // En caso de que body.ok sea null,
      // lanza un mensaje de error
      const error_msg = body.errors
        ? Object.values( body.errors )[0].msg
        : body.msg

      Swal.fire( 'Error', error_msg, 'error' );
    }
  }
};

export const startChecking = () => {
  return async ( dispatch ) => {
    const response = await fetchWithToken( "auth/renew" );
    const body = await response.json();

    if ( body.ok ) {
      // Se guarda el token en el storage
      localStorage.setItem('token', body.token);

      // Para saber cuando va a expirar el token
      localStorage.setItem( 'token-init-date', new Date().getTime() );

      // Dispara la acción del login
      dispatch(
        login({
          uid: body.uid,
          name: body.name
        })
      );
    } else {
      dispatch( checkingFinish() );
    }
  };
};

const checkingFinish = () => ({ type: types.authCheckingFinish });

const login = ( user ) => ({
  type: types.authLogin,
  payload: user
});

export const startLogout = () => {
  return ( dispatch ) => {

    // Borrar el localStorage
    localStorage.clear();

    dispatch( logout() );

  };
};

const logout = () => ({ type: types.authLogout });