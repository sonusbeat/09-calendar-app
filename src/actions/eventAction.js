import { fetchWithToken } from "../helpers/fetch";
import types from "../types/types";
import prepareEvents from '../helpers/prepareEvents';
import Swal from 'sweetalert2';

export const eventStartAddNew = ( event ) => {
  return async ( dispatch, getState ) => {
    const { uid, name } = getState().auth;

    try {

      const response = await fetchWithToken( "events", event, "POST" );
      const body = await response.json();

      if( body.ok ) {
        event.id = body.event.id;

        event.user = {
          _id: uid,
          name: name,
        };

        dispatch( eventAddNew( event ) );

        Swal.fire({
          position: 'center-center',
          icon: 'success',
          title: 'El evento ha sido creado',
          showConfirmButton: false,
          timer: 1500
        });
      }

    } catch(error) {

      console.log(error);

    }
  };
};

const eventAddNew = ( event ) => ({
  type: types.eventAddNew,
  payload: event
});

export const eventSetActive = ( event ) => ({
  type: types.eventSetActive,
  payload: event
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent
});

export const eventStartUpdate = ( event ) => {

  return async ( dispatch ) => {

    try {

      const response = await fetchWithToken( `events/${event.id}`, event, "PATCH" );
      const body = await response.json();

      if ( body.ok ) {
        dispatch( eventUpdated( event ) );

        Swal.fire({
          position: 'center-center',
          icon: 'success',
          title: 'El evento ha sido actualizado',
          showConfirmButton: false,
          timer: 1500
        });

      } else {
        Swal.fire("Error", body.msg, "error");
      }

    } catch (error) {
      console.log( error );
    }

  };
};

const eventUpdated = ( event ) => ({
  type: types.eventUpdated,
  payload: event
});

export const eventDeleted = () => ({ type: types.eventDeleted });


export const eventStartLoading = () => {
  return async ( dispatch ) => {

    try {

      const response = await fetchWithToken( "events" );
      const body = await response.json();

      const events = prepareEvents( body.events );

      if( body.ok ) {

        // Disparamos la acciÃ³n para enviar los eventos al state del redux
        dispatch( eventLoaded( events ) );

      }

    } catch (error) {
      console.log( error );
    }

    // dispatch(eventLoaded( events ));

  }
};

const eventLoaded = ( events ) => ({
  type: types.eventLoaded,
  payload: events,
});
