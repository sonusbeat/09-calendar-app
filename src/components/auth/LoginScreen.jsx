import "../../styles/main.scss";
import useForm from '../../hooks/useForm';
import { useDispatch } from 'react-redux';
import { startLogin, startRegister } from '../../actions/auth';
import Swal from 'sweetalert2';

const LoginScreen = () => {

  const dispatch = useDispatch();

  const [ formLoginValues, handleLoginInputChange ] = useForm({
    loginEmail:    "",
    loginPassword: "",
  });

  const { loginEmail, loginPassword } = formLoginValues;

  const [ formRegisterValues, handleRegisterInputChange ] = useForm({
    registerName:          "",
    registerEmail:         "",
    registerPassword:      "",
    passwordConfirmation:  "",
  });

  const { registerName, registerEmail, registerPassword, passwordConfirmation } = formRegisterValues;

  // Login de usuario
  const handleLogin = (event) => {
    event.preventDefault();

    dispatch( startLogin( loginEmail, loginPassword ) );
  };

  // Registro de usuario
  const handleRegister = (event) => {
    event.preventDefault();

    if( registerPassword !== passwordConfirmation ) {
      return Swal.fire("Error", "Las contrase√±as deben coincidir", "error");
    }

    dispatch( startRegister( registerName, registerEmail, registerPassword ) );
  };

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1 mb-5">
          <h3>Login</h3>

          <form onSubmit={ handleLogin }>
            <div className="form-group">
              <input
                name="loginEmail"
                value={ loginEmail }
                onChange={ handleLoginInputChange }
                type="text"
                className="form-control"
                placeholder="Correo"
                autoComplete="off"
              />
            </div>
            <div className="form-group mb-4">
              <input
                name="loginPassword"
                value={ loginPassword }
                onChange={ handleLoginInputChange }
                type="password"
                className="form-control"
                autoComplete="off"
              />
            </div>
            <div className="form-group text-center">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>

        </div>

        <div className="col-md-6 login-form-2 mb-5">
          <h3>Register</h3>

          <form onSubmit={ handleRegister }>
            <div className="form-group">
              <input
                name="registerName"
                type="text"
                className="form-control"
                placeholder="Name"
                autoComplete="off"
                onChange={ handleRegisterInputChange }
                value={ registerName }
              />
            </div>
            <div className="form-group">
              <input
                name="registerEmail"
                type="email"
                className="form-control"
                placeholder="Email"
                autoComplete="off"
                onChange={ handleRegisterInputChange }
                value={ registerEmail }
              />
            </div>
            <div className="form-group">
              <input
                name="registerPassword"
                type="password"
                className="form-control"
                placeholder="Password"
                autoComplete="off"
                onChange={ handleRegisterInputChange }
                value={ registerPassword }
              />
            </div>

            <div className="form-group mb-4">
              <input
                name="passwordConfirmation"
                type="password"
                className="form-control"
                placeholder="Password Confirmation"
                autoComplete="off"
                onChange={ handleRegisterInputChange }
                value={ passwordConfirmation }
              />
            </div>

            <div className="form-group text-center">
              <input type="submit" className="btnSubmit" value="Create Account" />
            </div>
          </form>

        </div>
      </div>
    </div>
  );
};

export default LoginScreen;
