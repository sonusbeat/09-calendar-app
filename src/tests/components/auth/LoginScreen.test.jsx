import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import LoginScreen from "../../../components/auth/LoginScreen";
import { startLogin } from "../../../actions/auth";

jest.mock("../../../actions/auth", () => ({
    startLogin: jest.fn(),
}));

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initialState = {};
const store = mockStore( initialState );

// Simular el dispatch para las pruebas
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <LoginScreen />
    </Provider>
);

describe("Pruebas en <LoginScreen />", () => {

    test("Deberia mostrarse correctamente", () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test("DeberÃ­a de llamar el dispatch del login", () => {

        const email    = "qbixmex@gmail.com";
        const password = "0123456789";

        wrapper.find('input[name="loginEmail"]').simulate("change", {
            target: {
                name : "loginEmail",
                value: email
            }
        });

        wrapper.find('input[name="loginPassword"]').simulate("change", {
            target: {
                name : "loginPassword",
                value: password
            }
        });

        wrapper.find("#login").prop("onSubmit")({
            preventDefault(){}
        });

        expect( startLogin ).toHaveBeenCalledWith( email, password );

    });

});
