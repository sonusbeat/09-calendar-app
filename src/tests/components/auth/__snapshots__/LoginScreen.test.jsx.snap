// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> Deberia mostrarse correctamente 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1 mb-5"
    >
      <h3>
        Login
      </h3>
      <form
        id="login"
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group mb-4"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group text-center"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2 mb-5"
    >
      <h3>
        Register
      </h3>
      <form
        id="register"
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="registerPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group mb-4"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="passwordConfirmation"
            onChange={[Function]}
            placeholder="Password Confirmation"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group text-center"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Create Account"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
